internal/model/metrics.go

```diff
@@ -0,0 +1,180 @@
+	DataEntry struct {
+		// The metric's name.
+		Name string
+		// The type of metric, i.e. which instrument is the data point for, one of [counter, gauge, histogram, summary].
+		Type InstrumentType
+		// The type of data source, one of [PROMETHEUS].
+		SourceType MetricsSourceType
+		// A freeform description of the metric, giving more detail on the data it represents.
+		Description string
+		// A list of data points that make up the entry.
+		Values []DataPoint
+	}
+
+	// DataPoint is a single data point for an entry. An entry can have multiple points.
+	DataPoint struct {
+		// Metric point name.
+		Name string
+		// Any potential labels associated with the data point.
+		Labels map[string]string
+		// The actual value of the data point.
+		Value any
+	}
+
+	// ExportType is a union type for all the data types we can convert to from a DataEntry.
+	ExportType interface {
+		metricdata.Metrics
+	}
+
+	// Converter is a conversion function from a DataEntry to the specified target type, which can then be exported.
+	Converter[T ExportType] func(input DataEntry) (T, error)
+
+	// Exporter acts as a data sink for metrics data generated by MetricsProducers.
+	Exporter interface {
+		// Export sends the given entry to the sink.
+		Export(entry DataEntry) error
+		// StartSink begins listening for Export() calls.
+		StartSink(ctx context.Context)
+	}
+
+	// A MetricsProducer will produce metrics from its configured data source
+	MetricsProducer interface {
+		// Produce collects metrics from the underlying data sources.
+		Produce(ctx context.Context) ([]DataEntry, error)
+		// The Type of the implementing sources.
+		Type() MetricsSourceType
+	}
+)
+
+func (de *DataEntry) ToBusMessage() *bus.Message {
+	return &bus.Message{
+		Topic: bus.MetricsTopic,
+		Data:  de,
+	}
+}
+
+type (
+	EntryBuilder struct {
+		entry DataEntry
+	}
+
+	BuilderOption func(eb *EntryBuilder)
+)
+
+func NewEntryBuilder(options ...BuilderOption) EntryBuilder {
+	builder := EntryBuilder{}
+
+	for _, opt := range options {
+		opt(&builder)
+	}
+
+	return builder
+}
+
+func WithSourceType(t MetricsSourceType) func(*EntryBuilder) {
+	return func(eb *EntryBuilder) {
+		eb.entry.SourceType = t
+	}
+}
+
+func (eb *EntryBuilder) WithName(n string) {
+	eb.entry.Name = n
+}
+
+func (eb *EntryBuilder) WithDescription(d string) {
+	eb.entry.Description = d
+}
+
+func (eb *EntryBuilder) WithType(t InstrumentType) {
+	eb.entry.Type = t
+}
+
+func (eb *EntryBuilder) WithSourceType(t MetricsSourceType) {
+	eb.entry.SourceType = t
+}
+
+func (eb *EntryBuilder) WithValues(dps ...DataPoint) {
+	eb.entry.Values = append(eb.entry.Values, dps...)
+}
+
+func (eb *EntryBuilder) Build() DataEntry {
+	return eb.entry
+}
+
+func (eb *EntryBuilder) CanBuild() bool {
+	return eb.entry.Name != "" && eb.entry.Description != "" && eb.entry.SourceType != 0 && eb.entry.Type != Unknown
+}
+
+type MetricsSourceType int
+
+const (
```

Comment:
ideally would like one type block and one const block with both grouped together near the top of the file. Seems to be a lot mixed around this file

Reply:
Yeah I'll group em' together 👍